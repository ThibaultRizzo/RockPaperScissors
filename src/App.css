#App {
  display: grid;
  grid-template-columns: 4fr 1fr;
  grid-template-rows: 1fr 3fr 2fr;
  grid-template-areas: 'header header' 'moves history' 'result history';
  grid-gap: 10px 10px;
  place-items: stretch stretch;
  height: 100%;
  /* Mandatory to get full viewport height */
}

#Header {
  background-color: #222;
  color: white;
  grid-area: header;
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  justify-content: space-around;
  padding: 10px;
}

#header-options {
  align-self: flex-start;
  margin: 5px 20px;
}

#MoveContainer {
  grid-area: moves;
  display: flex;
  flex-flow: row wrap;
  justify-content: space-around;
  align-items: center;
}

#GameResult {
  grid-area: result;
}

#PlayersHistory {
  grid-area: history;
}

header {
  font-size: 25px;
}

input.special-button {
  width: 120px;
  height: 40px;
  border: none;
  border-radius: 15px;
  box-shadow: white 0px 0px 2px 2px;
  background-color: white;
  color: black;
  font-size: 16px;
  transition: font-size 0.4s, box-shadow 0.4s, color 0.4s;
}

input.special-button:hover {
  background-color: rgba(255, 255, 255, 0.734);
  box-shadow: rgba(255, 255, 255, 0.734) 0px 0px 2px 10px;
  font-size: 18px;
  font-weight: bold;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

.full-page-background {
  background-color: rgba(0, 0, 0, 0.8);
  height: 100%;
  grid-column: header;
  grid-row: header / result;
  opacity: 1;

  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  overflow-x: hidden;
  z-index: 0;
  transition: all 0.5s;
}

.draw {
  color: blue;
}

.win {
  color: green;
}

.lose {
  color: red;
}

.deactivation-panel {
  grid-area: moves;
  background-color: white;
  opacity: 0.6;
  height: 50%;
  align-self: flex-end;
}

.smooth-appear {
  animation: appear 0.5s ease-in-out;
  animation-iteration-count: 1;
}

.jump {
  animation: jump 0.5s ease-in-out 0.3s infinite alternate;
}

@keyframes appear {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

@keyframes jump {
  from {
    bottom: 0px;
    position: relative;
  }
  to {
    bottom: 10px;
    position: relative;
  }
}
