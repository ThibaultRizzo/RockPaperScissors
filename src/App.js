import React, { Component } from 'react';
import './App.css';
import Overlay from './Overlay/Overlay.js';
import Header from './Header/Header.js';
import MoveContainer from './MoveContainer/MoveContainer.js';
import GameResult from './GameResult/GameResult.js';
import ModalPopup from './ModalPopup.js';
import PlayersHistory from './PlayersHistory/PlayersHistory.js';
import { resolveTurn, summarizeGame, GameType } from './Rules.js';

class App extends Component {
  constructor(props) {
    super(props);
    this.onPlayerMove = this.onPlayerMove.bind(this);
    this.setGameType = this.setGameType.bind(this);
    this.displayResults = this.displayResults.bind(this);
    this.closePopup = this.closePopup.bind(this);
    this.incrementturnCount = this.incrementturnCount.bind(this);
    this.onGameEnd = this.onGameEnd.bind(this);
    this.addGameResults = this.addGameResults.bind(this);
    this.state = {
      turn: null,
      gameType: null,
      showModal: false,
      currentturnCount: false,
      maxturnCount: false,
      isGameFinished: false,
      turns: [],
      games: []
    };
  }

  /**
   * Updates state according to the clicked option
   */
  onPlayerMove(move) {
    let turn = resolveTurn(move);
    this.displayResults(turn);
  }

  /**
   * Historizes current game results with details
   * Details are: 
   *  - result: game output from P1 perspective
   *  - scoreP1: how many wins for P1
   *  - scoreP2:  how many wins for P2
   *  - gameType: 'Player vs AI' or 'AI vs AI'
   * @param {*} game 
   */
  onGameEnd(game) {
    let summarizedGame = this.addGameResults(game);
    this.setState(prevState => ({
      games: [...prevState.games, summarizedGame]
    }));
  }

  /**
   * Enhances game object with details property.
   * ex:
   *      details: {
   *          result: -> Result enum object
   *          scoreP1: -> Number of turn won by P1
   *          scoreP2: -> Number of turn won by P2
   *      }
   * @param {*} game
   */
  addGameResults(game) {
    let gameResult = summarizeGame(game, this.state.gameType);
    return [...game, gameResult];
  }

  /**
   * Increment turn number if less than maximum turn count else returns -1.
   */
  incrementturnCount() {
    let currentturnCount = this.state.currentturnCount;
    let maxturnCount = this.state.maxturnCount;
    return currentturnCount < maxturnCount ? currentturnCount + 1 : -1;
  }

  /**
   * Updates the state given current play and assess wether or not game is finished
   * @param {*} currentTurn 
   */
  displayResults(currentTurn) {
    let gameStatus = false;
    let updatedTurnArray = this.state.turns;
    updatedTurnArray.push(currentTurn);
    let nextturnCount = this.incrementturnCount();
    if (nextturnCount === this.state.maxturnCount) {
      gameStatus = true;
      this.onGameEnd(this.state.turns);
    }
    // If turnCount gets out of boundaries, state is not updated
    if (nextturnCount !== -1) {
      this.setState({
        turn: currentTurn,
        showModal: true,
        isGameFinished: gameStatus,
        currentturnCount: nextturnCount,
        turns: updatedTurnArray
      });
    }
  }

  /**
   * Reset new game and add game type
   * @param {*} turnNumber 
   * @param {*} type 
   */
  setGameType(turnCount, type) {
    this.setState({
      gameType: type,
      currentturnCount: 0,
      maxturnCount: turnCount,
      isGameFinished: false,
      turns: []
    });
  }

  /**
   * Close modal popup
   */
  closePopup() {
    this.setState({ showModal: false });
  }

  render() {
    let content;
    let overlay = false;
    let modal = false;
    let gameStatus = this.state.isGameFinished;
    if (
      this.state.showModal &&
      this.state.turn &&
      this.state.gameType === GameType.PlayerVsAI
    ) {
      modal = (
        <ModalPopup
          closePopup={this.closePopup}
          result={this.state.turn.result}
        />
      );
    }
    if (!this.state.gameType) {
      overlay = <Overlay onOptionChosed={this.setGameType} />;
    }
    content = (
      <React.StrictMode>
        <Header
          isGameFinished={gameStatus}
          onCreateNewGame={this.setGameType}
        />
        <MoveContainer
          isGameFinished={gameStatus}
          maxturnCount={this.state.maxturnCount}
          gameType={this.state.gameType}
          onMoveClicked={this.onPlayerMove}
        />
        <div
          className={this.state.isGameFinished ? 'deactivation-panel' : ''}
        />
        <GameResult
          isGameFinished={gameStatus}
          turnArray={this.state.turns}
          turnCount={this.state.currentturnCount}
        />
        <PlayersHistory gameHistory={this.state.games} />
      </React.StrictMode>
    );
    return (
      <div id="App">
        {overlay}
        {content}
        {modal}
      </div>
    );
  }
}

export default App;
