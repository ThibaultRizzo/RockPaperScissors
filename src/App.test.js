import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const clicked = jest.fn();
  const rendered = renderer.create(<App />);
  expect(rendered.toJSON()).toMatchSnapshot();
});
