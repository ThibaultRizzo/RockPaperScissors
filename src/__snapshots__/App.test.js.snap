// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id="App"
>
  <div
    className="full-page-background smooth-appear"
    id="Overlay"
  >
    <input
      className="overlay-button special-button"
      onClick={[Function]}
      type="button"
      value="Player vs AI"
    />
    <input
      className="overlay-button special-button"
      onClick={[Function]}
      type="button"
      value="AI vs AI"
    />
    <div
      className="turn-dropdown"
    >
      <input
        className="selected-dropdown-value dropdown-value"
        onClick={[Function]}
        type="button"
        value={3}
      />
      <div
        className="dropdown-content"
      />
    </div>
  </div>
  <div
    id="Header"
  >
    <header>
      Waste an Hour Having Fun
    </header>
    <div
      id="header-options"
    >
      <input
        className="special-button "
        onClick={[Function]}
        type="button"
        value="New Game"
      />
    </div>
  </div>
  <div
    id="MoveContainer"
  >
    <h1
      className="game-type"
    />
    <h2
      className="number-turn"
    >
      Number of turns: 
    </h2>
    <div
      className="Move  "
    >
      <input
        name="ROCK"
        onClick={[Function]}
        type="button"
      />
    </div>
    <div
      className="Move  "
    >
      <input
        name="PAPER"
        onClick={[Function]}
        type="button"
      />
    </div>
    <div
      className="Move  "
    >
      <input
        name="SCISSORS"
        onClick={[Function]}
        type="button"
      />
    </div>
  </div>
  <div
    className=""
  />
  <div
    id="GameResult"
  >
    <div
      className="result-grid"
    />
  </div>
  <div
    id="PlayersHistory"
  >
    <h1>
      Players History
    </h1>
    <div
      className="win-rate"
    >
      Win Rate: 
      0%
    </div>
  </div>
</div>
`;
